### PR TITLE
chore: Improves user agent in the provider to include Terraform CLI version

### PR DESCRIPTION
**Pull request by maastha**
## Description

Improves user agent in the provider to include Terraform CLI version

Link to any related issue(s): [CLOUDP-237005](https://jira.mongodb.org/browse/CLOUDP-237005)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [X] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Further comments

TF Plugin Framework based resource API call logs:
```
2024/04/05 15:17:50 
GET /api/atlas/v2/groups/6602bedbc32461337d92e2a5/settings HTTP/1.1
Host: cloud-dev.mongodb.com
User-Agent: terraform-provider-mongodbatlas/devel Terraform/1.7.0-rc1
Accept: application/vnd.atlas.2023-01-01+json
Accept-Encoding: gzip
```

TF Plugin SDK based resource API call logs:
```
2024/04/05 15:17:52 
POST /api/atlas/v2/groups/6602d0208210bb7b10f5f8a8/clusters HTTP/1.1
Host: cloud-dev.mongodb.com
User-Agent: terraform-provider-mongodbatlas/devel Terraform/1.7.0-rc1
Content-Length: 302
Accept: application/vnd.atlas.2023-02-01+json
```

### CODE REVIEW

File: internal/config/client.go [hunk 1]
```diff
@@ -175,3 +183,18 @@ func (c *MongoDBClient) GetRealmClient(ctx context.Context) (*realm.Client, erro
...
+		if currentUA != "" {
+			currentUA += " "
+		}
+		currentUA += additionalInfo
+	}
+	c.UserAgent = currentUA
+}
+
+func TerraformVersionUserAgentInfo(tfVersion string) string {
+	return fmt.Sprintf("Terraform/%s", tfVersion)
```
Comment (oarbusi) [inline internal/config/client.go]: small nit: create and use const like `toolName`
```
return fmt.Sprintf("%s/%s", terraformName, tfVersion)
```

**Review by oarbusi — APPROVED** [review]
Great work! LGTM

File: internal/provider/provider_sdk2.go [hunk 2]
```diff
@@ -133,11 +133,15 @@ func NewSdkV2Provider(proxyPort *int) *schema.Provider {
...
-		ResourcesMap:         getResourcesMap(),
-		ConfigureContextFunc: providerConfigure(proxyPort),
+		DataSourcesMap: getDataSourcesMap(),
+		ResourcesMap:   getResourcesMap(),
 	}
 	addPreviewFeatures(provider)
+
+	provider.ConfigureContextFunc = func(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
+		return providerConfigure(provider, proxyPort)(ctx, d)
+	}
```
Comment (AgustinBettati) [inline internal/provider/provider_sdk2.go]:
```suggestion
	provider.ConfigureContextFunc = providerConfigure(provider, proxyPort)
```

File: internal/provider/provider_sdk2.go [hunk 3]
```diff
@@ -296,6 +300,7 @@ func providerConfigure(proxyPort *int) func(ctx context.Context, d *schema.Resou
 			BaseURL:      d.Get("base_url").(string),
 			RealmBaseURL: d.Get("realm_base_url").(string),
 			ProxyPort:    proxyPort,
+			UserAgent:    config.TerraformVersionUserAgentInfo(provider.TerraformVersion),
```
Comment (AgustinBettati) [inline internal/provider/provider_sdk2.go]: a little hard to follow, but I assume that when ConfigureContextFunc is called it mutates the `TerraformVersion` property of the schema.Provider?

File: internal/provider/provider.go [hunk 4]
```diff
@@ -231,6 +231,7 @@ func (p *MongodbtlasProvider) Configure(ctx context.Context, req provider.Config
 		PrivateKey:   data.PrivateKey.ValueString(),
 		BaseURL:      data.BaseURL.ValueString(),
 		RealmBaseURL: data.RealmBaseURL.ValueString(),
+		UserAgent:    config.TerraformVersionUserAgentInfo(req.TerraformVersion),
```
Comment (AgustinBettati) [inline internal/provider/provider.go]: [q] Could both the provider version and the terraform version be defined in a single place so it is easy to see the complete user agent format?

File: internal/config/client.go [hunk 5]
```diff
@@ -175,3 +183,18 @@ func (c *MongoDBClient) GetRealmClient(ctx context.Context) (*realm.Client, erro
...
+	if !strings.Contains(currentUA, additionalInfo) {
+		if currentUA != "" {
+			currentUA += " "
+		}
+		currentUA += additionalInfo
+	}
+	c.UserAgent = currentUA
+}
+
+func TerraformVersionUserAgentInfo(tfVersion string) string {
```
Comment (AgustinBettati) [inline internal/config/client.go]: Curious if from the analytics team there is any preference in how this terraform information is added so we don't impact any existing queries that make use of the user agent. Maybe worth sending a message to Victor to verify.

**Review by AgustinBettati — APPROVED** [review]
LGTM, just some follow up doubts

File: internal/provider/provider.go (hunk 4 above)
Comment (maastha) [reply internal/provider/provider.go]: it can be but I have kept it separate. Ideally user agent should be taken care of when creating new client (like we already do today) but in this case I had to set the new Terraform version user agent in the Provider because that's only where it's available.
Defining both in single place would mean we take provider version out of the client creation which I don't think we should do

File: internal/provider/provider_sdk2.go (hunk 3 above)
Comment (maastha) [reply internal/provider/provider_sdk2.go]: right

File: internal/config/client.go (hunk 5 above)
Comment (maastha) [reply internal/config/client.go]: Looks like Victor is out until after next week, I think let's merge this for now and in the meantime I'll send a ping to him anyway so we can make any updates later if required

File: internal/provider/provider.go (hunk 4 above)
Comment (AgustinBettati) [reply internal/provider/provider.go]: I see, one option could be to define a property TerraformVersion in the Config struct (instead of UserAgent), and then when creating a new client define the user agent with both the terraform and provider versions available. Let me know if you feel this makes it clearer or not.

File: internal/config/client.go (hunk 5 above)
Comment (maastha) [reply internal/config/client.go]: Reached out to Victor/natalya over Slack, included you both in the conversation so he can respond whenever he is back

File: internal/config/client.go [hunk 6]
```diff
@@ -6,25 +6,28 @@ import (
...
+	"github.com/mongodb/terraform-provider-mongodbatlas/version"
 )
 
 const (
 	toolName = "terraform-provider-mongodbatlas"
 )
 
 var (
-	userAgent = fmt.Sprintf("%s/%s", toolName, version.ProviderVersion)
+	userAgentProviderVersion = fmt.Sprintf("%s/%s", toolName, version.ProviderVersion)
```
Comment (lantoli) [inline internal/config/client.go]: would it be useful to add some tests in the PR?

**Review by EspenAlbert — APPROVED** [review]

File: internal/config/client.go [hunk 7]
```diff
@@ -175,3 +183,18 @@ func (c *MongoDBClient) GetRealmClient(ctx context.Context) (*realm.Client, erro
 
 	return realmClient, nil
 }
+
+func (c *Config) appendToUserAgent(additionalInfo string) {
+	currentUA := c.UserAgent
+	if !strings.Contains(currentUA, additionalInfo) {
```
Comment (marcosuma) [inline internal/config/client.go]: why do we need to check if it's already there? is the code being called more than once?

File: internal/provider/provider.go (hunk 4 above)
Comment (marcosuma) [reply internal/provider/provider.go]: I was thinking the same thing after spending some time reviewing this. Having `TerraformVersion` would help a lot in terms of readability. In fact during client creation you can "compose" the user agent in-place without even having to have a special append function IMO. 

Also this solution would help to handle logs / problems: e.g. if `TerraformVersion` is empty during the client creation phase then we should handle it with a warning probably since it's something that needs attention

File: internal/config/client.go (hunk 7 above)
Comment (maastha) [reply internal/config/client.go]: >is the code being called more than once?

no, but was trying to keep the method generic

File: internal/provider/provider.go (hunk 4 above)
Comment (maastha) [reply internal/provider/provider.go]: I have added TerraformVersion in the Config struct, let me know what you think

**Review by AgustinBettati — APPROVED** [review]
LGTM, just to confirm the conclusion with analytics team was to have the provider version first and then the terraform version?

**Comment by maastha** [comment]
>LGTM, just to confirm the conclusion with analytics team was to have the provider version first and then the terraform version?

This was Gus's suggestion, yes. I have updated the testing example in the PR accordingly.

**Review by marcosuma — APPROVED** [review]
Awesome job!
